test/compress/harmony.js

```diff
@@ -804,3 +804,39 @@ object_spread_of_sequence: {
         console.log({ ...o || o });
     }
 }
+
+// issue 2316
+class_name_can_be_preserved_with_reserved: {
+    mangle = {
+        reserved: ['Foo']
+    }
+    input: {
+        function x() {
+            class Foo {};
+            Foo.bar;
+            class Bar {};
+            Bar.foo;
+        }
+
+        function y() {
+            var Foo = class Foo {};
+            Foo.bar();
+            var Bar = class Bar {};
+            Bar.bar();
```

Comment:
Thanks.

You're very quick - you caught my suggestion just before an edit. I had changed it to match the pattern of properties for `function x` above:
```
        function y() {
             var Foo = class Foo {};
             Foo.bar;
             var Bar = class Bar {};
             Bar.foo;
        }

Reply:
@alexlamsl No need to hold up this PR for this.